File: scala/src/test/scala/sodium/StreamTester.scala
package sodium

import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Test

import scala.collection.mutable.ListBuffer

class StreamTester {
  @After
  def tearDown(): Unit = {
    System.gc()
    Thread.sleep(100)
  }

  @Test
  def testSendStream(): Unit = {
    val e = new StreamSink[Int]()
    val out = new ListBuffer[Int]()
    val l = e.listen(out.+=)
    e.send(5)
    l.unlisten()
    assertEquals(List(5), out)
    e.send(6)
    assertEquals(List(5), out)
  }

  @Test
  def testMap(): Unit = {
    val e = new StreamSink[Int]()
    val m = e.map(x => x.toString)
    val out = new ListBuffer[String]()
    val l = m.listen(out.+=)
    e.send(5)
    l.unlisten()
    assertEquals(List("5"), out)
  }

  @Test
  def testMapTo(): Unit = {
    val e = new StreamSink[Int]()
    val m = e.mapTo("fusebox")
    val out = new ListBuffer[String]()
    val l = m.listen(out.+=)
    e.send(5)
    e.send(6)
    l.unlisten()
    assertEquals(List("fusebox", "fusebox"), out)
  }

  @Test
  def testMergeNonSimultaneous(): Unit = {
    val e1 = new StreamSink[Int]()
    val e2 = new StreamSink[Int]()
    val out = new ListBuffer[Int]()
    val l = e2.orElse(e1).listen(out.+=)
    e1.send(7)
    e2.send(9)
    e1.send(8)
    l.unlisten()
    assertEquals(List(7, 9, 8), out)
  }

  @Test
  def testMergeSimultaneous(): Unit = {
    val s1 = new StreamSink[Int]((_, r) => r)
    val s2 = new StreamSink[Int]((_, r) => r)
    val out = new ListBuffer[Int]()
    val l = s2.orElse(s1).listen(out.+=)
    Transaction(_ => {
      s1.send(7)
      s2.send(60)
    })
    Transaction(_ => s1.send(9))
    Transaction(_ => {
      s1.send(7)
      s1.send(60)
      s2.send(8)
      s2.send(90)
    })
    Transaction(_ => {
      s2.send(8)
      s2.send(90)
      s1.send(7)
      s1.send(60)
    })
    Transaction(_ => {
      s2.send(8)
      s1.send(7)
      s2.send(90)
      s1.send(60)
    })
    l.unlisten()
    assertEquals(List(60, 9, 90, 90, 90), out)
  }

  @Test
  def testCoalesce(): Unit = {
    val s = new StreamSink[Int]((a, b) => a + b)
    val out = new ListBuffer[Int]()
    val l = s.listen(out.+=)
    Transaction(_ => s.send(2))
    Transaction(_ => {
      s.send(8)
      s.send(40)
    })
    l.unlisten()
    assertEquals(List(2, 48), out)
  }

  @Test
  def testFilter(): Unit = {
    val e = new StreamSink[Char]()
    val out = new ListBuffer[Char]()
    val l = e.filter(c => Character.isUpperCase(c)).listen(c => out.+=(c))
    List('H', 'o', 'I').foreach(e.send(_))
    l.unlisten()
    assertEquals(List('H', 'I'), out)
  }

  @Test
  def testFilterOptional(): Unit = {
    val e = new StreamSink[Option[String]]()
    val out = new ListBuffer[String]()
    val l = Stream.filterOptional(e).listen(s => out.+=(s))
    List(Some("tomato"), None, Some("peach")).foreach(e.send(_))
    l.unlisten()
    assertEquals(List("tomato", "peach"), out)
  }

  @Test
  def testLoopStream(): Unit = {
    val ea = new StreamSink[Int]()
    val ec = Transaction(_ => {
      val eb = new StreamLoop[Int]()
      val ec_ = ea.map(x => x % 10).merge(eb, (x, y) => x + y)
      val eb_out = ea.map(x => x / 10).filter(x => x != 0)
      eb.loop(eb_out)
      ec_
    })
    val out = new ListBuffer[Int]()
    val l = ec.listen(out.+=)
    List(2, 52).foreach(ea.send(_))
    l.unlisten()
    assertEquals(List(2, 7), out)
  }

  @Test
  def testGate(): Unit = {
    val ec = new StreamSink[Char]()
    val epred = new BehaviorSink(true)
    val out = new ListBuffer[Char]()
    val l = ec.gate(epred).listen(out.+=)
    ec.send('H')
    epred.send(false)
    ec.send('O')
    epred.send(true)
    ec.send('I')
    l.unlisten()
    assertEquals(List('H', 'I'), out)
  }

  @Test
  def testCollect(): Unit = {
    val ea = new StreamSink[Int]()
    val out = new ListBuffer[Int]()
    val sum = ea.collect[Int, Int](0, (a, s) => (a + s + 100, a + s))
    val l = sum.listen(out.+=)
    List(5, 7, 1, 2, 3).foreach(ea.send(_))
    l.unlisten()
    assertEquals(List(105, 112, 113, 115, 118), out)
  }

  @Test
  def testAccum(): Unit = {
    val ea = new StreamSink[Int]()
    val out = new ListBuffer[Int]()
    val sum = ea.accum[Int](100, (a, s) => a + s)
    val l = sum.listen(out.+=)
    List(5, 7, 1, 2, 3).foreach(ea.send(_))
    l.unlisten()
    assertEquals(List(100, 105, 112, 113, 115, 118), out)
  }

  @Test
  def testOnce(): Unit = {
    val e = new StreamSink[Char]()
    val out = new ListBuffer[Char]()
    val l = e.once().listen(out.+=)
    List('A', 'B', 'C').foreach(e.send(_))
    l.unlisten()
    assertEquals(List('A'), out)
  }

  @Test
  def testDefer(): Unit = {
    val e = new StreamSink[Char]()
    val b = e.hold(' ')
    val out = new ListBuffer[Char]()
    val l = Operational.defer(e).snapshot(b).listen(out.+=)
    List('C', 'B', 'A').foreach(e.send(_))
    l.unlisten()
    assertEquals(List('C', 'B', 'A'), out)
  }

  @Test
  def testStreamLoop(): Unit = {
    val ss = new StreamSink[Int]()
    val s = Transaction(_ => {
      val sl = new StreamLoop[Int]()
      val c = sl.map(v => v + 2).hold(0)
      val s2 = ss.snapshot(c, (x: Int, y: Int) => x + y)
      sl.loop(s2)
      s2
    })
    val out = new ListBuffer[Int]()
    val l = s.listen(out.+=)
    List(3, 4, 7, 8).foreach(ss.send(_))
    l.unlisten()
    assertEquals(List(3, 9, 18, 28), out)
  }

  @Test
  def testStreamLoopDefer(): Unit = {
    val ss = new StreamSink[Int]()
    val s = Transaction(_ => {
      val sl = new StreamLoop[Int]()
      val streamLocal = Operational.defer(ss.orElse(sl).filter(v => v < 5).map(v => v + 1))
      sl.loop(streamLocal)
      sl
    })
    val out = new ListBuffer[Int]()
    val l = s.listen(out.+=)
    ss.send(2)
    l.unlisten()
    assertEquals(List(3, 4, 5), out)
  }

}

File: java/tests/java8/sodium/StreamTester.java
package sodium;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

import junit.framework.TestCase;

public class StreamTester extends TestCase {
	@Override
	protected void tearDown() throws Exception {
		System.gc();
		Thread.sleep(100);
	}

	public void testSendStream()
    {
        StreamSink<Integer> e = new StreamSink();
        List<Integer> out = new ArrayList();
        Listener l = e.listen(x -> { out.add(x); });
        e.send(5);
        l.unlisten();
        assertEquals(Arrays.asList(5), out);
        e.send(6);
        assertEquals(Arrays.asList(5), out);
    }

	public void testMap()
    {
        StreamSink<Integer> e = new StreamSink();
        Stream<String> m = e.map(x -> Integer.toString(x));
        List<String> out = new ArrayList();
        Listener l = m.listen((String x) -> { out.add(x); });
        e.send(5);
        l.unlisten();
        assertEquals(Arrays.asList("5"), out);
    }

    public void testMergeNonSimultaneous()
    {
        StreamSink<Integer> e1 = new StreamSink();
        StreamSink<Integer> e2 = new StreamSink();
        List<Integer> out = new ArrayList();
        Listener l = e1.merge(e2).listen(x -> { out.add(x); });
        e1.send(7);
        e2.send(9);
        e1.send(8);
        l.unlisten();
        assertEquals(Arrays.asList(7,9,8), out);
    }

    public void testMergeSimultaneous()
    {
        StreamSink<Integer> e = new StreamSink();
        List<Integer> out = new ArrayList();
        Listener l = e.merge(e).listen(x -> { out.add(x); });
        e.send(7);
        e.send(9);
        l.unlisten();
        assertEquals(Arrays.asList(7,7,9,9), out);
    }

    public void testMergeLeftBias()
    {
        StreamSink<String> e1 = new StreamSink();
        StreamSink<String> e2 = new StreamSink();
        List<String> out = new ArrayList();
        Listener l = e1.merge(e2).listen(x -> { out.add(x); });
        Transaction.runVoid(() -> {
            e1.send("left1a");
            e1.send("left1b");
            e2.send("right1a");
            e2.send("right1b");
        });
        Transaction.runVoid(() -> {
            e2.send("right2a");
            e2.send("right2b");
            e1.send("left2a");
            e1.send("left2b");
        });
        l.unlisten();
        assertEquals(Arrays.asList(
            "left1a", "left1b",
            "right1a", "right1b",
            "left2a", "left2b",
            "right2a", "right2b"
        ), out);
    }

    public void testCoalesce()
    {
        StreamSink<Integer> e1 = new StreamSink();
        StreamSink<Integer> e2 = new StreamSink();
        List<Integer> out = new ArrayList();
        Listener l =
             e1.merge(e1.map(x -> x * 100).merge(e2))
            .coalesce((Integer a, Integer b) -> a+b)
            .listen((Integer x) -> { out.add(x); });
        e1.send(2);
        e1.send(8);
        e2.send(40);
        l.unlisten();
        assertEquals(Arrays.asList(202, 808, 40), out);
    }
    
    public void testFilter()
    {
        StreamSink<Character> e = new StreamSink();
        List<Character> out = new ArrayList();
        Listener l = e.filter((Character c) -> Character.isUpperCase(c)).listen((Character c) -> { out.add(c); });
        e.send('H');
        e.send('o');
        e.send('I');
        l.unlisten();
        assertEquals(Arrays.asList('H','I'), out);
    }

    public void testFilterNotNull()
    {
        StreamSink<String> e = new StreamSink();
        List<String> out = new ArrayList();
        Listener l = e.filterNotNull().listen(s -> { out.add(s); });
        e.send("tomato");
        e.send(null);
        e.send("peach");
        l.unlisten();
        assertEquals(Arrays.asList("tomato","peach"), out);
    }

    public void testFilterOptional()
    {
        StreamSink<Optional<String>> e = new StreamSink();
        List<String> out = new ArrayList();
        Listener l = Stream.filterOptional(e).listen(s -> { out.add(s); });
        e.send(Optional.of("tomato"));
        e.send(Optional.empty());
        e.send(Optional.of("peach"));
        l.unlisten();
        assertEquals(Arrays.asList("tomato","peach"), out);
    }

    public void testLoopStream()
    {
        final StreamSink<Integer> ea = new StreamSink();
        Stream<Integer> ec = Transaction.<Stream<Integer>>run(() -> {
            StreamLoop<Integer> eb = new StreamLoop<Integer>();
            Stream<Integer> ec_ = ea.map(x -> x % 10).merge(eb, (x, y) -> x+y);
            Stream<Integer> eb_out = ea.map(x -> x / 10).filter(x -> x != 0);
            eb.loop(eb_out);
            return ec_;
        });
        List<Integer> out = new ArrayList();
        Listener l = ec.listen(x -> { out.add(x); });
        ea.send(2);
        ea.send(52);
        l.unlisten();
        assertEquals(Arrays.asList(2,7), out);
    }

    public void testGate()
    {
        StreamSink<Character> ec = new StreamSink();
        CellSink<Boolean> epred = new CellSink(true);
        List<Character> out = new ArrayList();
        Listener l = ec.gate(epred).listen(x -> { out.add(x); });
        ec.send('H');
        epred.send(false);
        ec.send('O');
        epred.send(true);
        ec.send('I');
        l.unlisten();
        assertEquals(Arrays.asList('H','I'), out);
    }

    public void testCollect()
    {
        StreamSink<Integer> ea = new StreamSink();
        List<Integer> out = new ArrayList();
        Stream<Integer> sum = ea.collect(100,
            //(a,s) -> new Tuple2(a+s, a+s)
            new Lambda2<Integer, Integer, Tuple2<Integer,Integer>>() {
                public Tuple2<Integer,Integer> apply(Integer a, Integer s) {
                    return new Tuple2<Integer,Integer>(a+s, a+s);
                }
            }
        );
        Listener l = sum.listen((x) -> { out.add(x); });
        ea.send(5);
        ea.send(7);
        ea.send(1);
        ea.send(2);
        ea.send(3);
        l.unlisten();
        assertEquals(Arrays.asList(105,112,113,115,118), out);
    }

    public void testAccum()
    {
        StreamSink<Integer> ea = new StreamSink();
        List<Integer> out = new ArrayList();
        Cell<Integer> sum = ea.accum(100, (a,s)->a+s);
        Listener l = sum.updates().listen((x) -> { out.add(x); });
        ea.send(5);
        ea.send(7);
        ea.send(1);
        ea.send(2);
        ea.send(3);
        l.unlisten();
        assertEquals(Arrays.asList(105,112,113,115,118), out);
    }

    public void testOnce()
    {
        StreamSink<Character> e = new StreamSink();
        List<Character> out = new ArrayList();
        Listener l = e.once().listen((x) -> { out.add(x); });
        e.send('A');
        e.send('B');
        e.send('C');
        l.unlisten();
        assertEquals(Arrays.asList('A'), out);
    }

    public void testDelay()
    {
        StreamSink<Character> e = new StreamSink();
        Cell<Character> b = e.hold(' ');
        List<Character> out = new ArrayList();
        Listener l = e.delay().snapshot(b).listen((x) -> { out.add(x); });
        e.send('C');
        e.send('B');
        e.send('A');
        l.unlisten();
        assertEquals(Arrays.asList('C','B','A'), out);
    }
}

